pin/gc/gc.go

```diff
@@ -0,0 +1,102 @@
+package gc
+
+import (
+	bstore "github.com/ipfs/go-ipfs/blocks/blockstore"
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	bserv "github.com/ipfs/go-ipfs/blockservice"
+	offline "github.com/ipfs/go-ipfs/exchange/offline"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	pin "github.com/ipfs/go-ipfs/pin"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	eventlog "github.com/ipfs/go-ipfs/thirdparty/eventlog"
+)
+
+var log = eventlog.Logger("gc")
+
+// GC performs a mark and sweep garbage collection of the blocks in the blockstore
+// first, it creates a 'marked' set and adds to it the following:
+// - all recursively pinned blocks, plus all of their descendants (recursively)
+// - all directly pinned blocks
+// - all blocks utilized internally by the pinner
+//
+// The routine then iterates over every block in the blockstore and
+// deletes any block that is not found in the marked set.
+func GC(ctx context.Context, bs bstore.GCBlockstore, pn pin.Pinner) (<-chan key.Key, error) {
```

Comment:
i think the "mark and sweep" functionality of this code should be abstracted into a function in `dag` pkg.

```go
Color(all <-chan key.Key, roots []key.Key) (colored, other key.KeySet) {
  colored := Descendants(roots)
  other := Complement(all, colored)
  return colored, other
}

func Complement(all <-chan key.Key, set key.KeySet) key.KeySet {
  c := key.NewKeySet()
  for key := range all {
    if !set.Has(key) { 
      c.Add(key)
    }
  }
  return c
}

func Descendants(roots []key.Key) key.KeySet {
  set := key.NewKeySet()
  for _, root in roots {
    set.Add(root)
    // add all descendants of root to colored
    dag.EnumerateChildren(...)
  }    
}
```

so that it can be used in other things, including listing the `indirect` pins in the `ipfs pin ls` cmd, and elsewhere when we mark roots to `provide` or `serve` recursively.